<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta name="author" content="Matthew Preston">
  <meta name="description" content="">

  <title>Mega.py Alignment Viewer</title>

  <!--CSS-->
  <link rel="stylesheet" href="../../css/bootstrap.min.css">
  <link rel="stylesheet" href="../../css/octicons.min.css">
  <link rel="stylesheet" href="../../css/article.css">
  <link rel="stylesheet" href="../../css/article-python.css">

  <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
  <![endif]-->
</head>
<body data-spy="scroll" data-target="#spy-target" data-offset="60">
  <!--Navbar-->
  <nav class="navbar navbar-default">
    <div class="container">
      <div class="navbar-header">
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#menuNavbar">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../../index.html">Matt Preston</a>
      </div>
      <div class="collapse navbar-collapse" id="menuNavbar">
        <ul class="nav navbar-nav navbar-right">
          <li><a href="https://github.com/matthewpreston/matthewpreston.github.io/tree/master/programs">Programs</a></li>
          <li><a href="../../Resume_Matthew_Preston.pdf" target="_blank">Resume</a></li>
          <li><a href="../../about.html">About</a></li>
        </ul>
      </div>
    </div>
  </nav>

  <div class="row wrapper">
    <!--Main container-->
    <article class="col-sm-9">
      <!--Title-->
      <div class="title">
        <h1>Mega.py</h1>
        <div class="small-divider"></div>
        <h4>Alignment Viewer</h4>
      </div>
      <!--Create double border effect-->
      <div class="outer-border">
        <div class="inner-border">
          <!--Presentation Image-->
          <div class="main-img">
            <a href="../../img/MegaAlignment.png"><img src="../../img/MegaAlignment.png" alt="Mega Alignment"></a>
          </div>
          <!--Introduction-->
          <div class="container-fluid" id="introduction">
            <h3>Introduction</h3>
            <div class="divider"></div>
            <p>&emsp;&emsp;&emsp;Written in Python, Mega.py is a complement to the popular aligner Mega(link). While Mega has a plethora of other varying functions (Muscle and ClustalW alignment algorithms, phylogeny construction, calculating evolutionary rates, etc.), this script solely focuses on displaying DNA and protein alignments in an user friendly manner. The motivation to write this program was to manage DNA ambiguity characters when translating to amino acid residues, which Mega cannot handle. The convenience of this allows the user to effortlessly visualize the resulting protein sequence to make inferences on its overall function or go back and decipher the intended nitrogenous base.</p>
          </div>
          <!--Features-->
          <div class="container-fluid" id="features">
            <h3>Features</h3>
            <div class="divider"></div>
            <ul>
              <li><strong>3 distinct output modes:</strong>
                <ul>
                  <li><strong>Reference:</strong> Uses 1st sequence in a FASTA file as a reference and pair-wise aligns the remaining sequences to it. Matching bases between the two strings are highlighted. Additionally, the sequences are translated at every possible frameshift position, allowing for forward open reading frames to be read off.
                  </li>
                  <li><strong>DNA:</strong> Clusters nucleotide sequences according to gene and displays the similiarities and positions between the grouping.
                  </li>
                  <li><strong>Protein:</strong> Clusters nucleotide sequences according to gene, translates each one to generate a protein sequence according to a suitable codon table, and displays the similarities and positions between the grouping.
                  </li>
                </ul>
              </li>
              <li>Robust against ambiguous nucleotides and amino acids; features a drop-down box to showcase each possible residue generated by enigmatic codons.</li>
              <li>A customizable colour palette for both DNA and protein alignments</li>
            </ul>
          </div>
          <!--Download/Github-->
          <div class="container-fluid" id="download">
            <h3>Download</h3>
            <div class="divider"></div>
            <table>
              <tr>
                <td><a href="MegaV1.0.zip" title="Download Mega.zip" class="button"><span class="glyphicon glyphicon-compressed"></span> Mega.zip</a></td>
                <td><a href="MegaV1.0.tar.gz" title="Download Mega.tar.gz" class="button"><span class="glyphicon glyphicon-compressed"></span> Mega.tar.gz</a></td>
                <td><a href="https://github.com/matthewpreston/matthewpreston.github.io/tree/master/programs/Mega" title="View on GitHub" class="button"><span class="octicon octicon-git-branch"></span>View on GitHub</a></td>
              </tr>
            </table>
          </div>
          <!--Usage/Demonstration-->
          <div class="container-fluid" id="demonstration">
            <h3>Demonstration</h3>
            <div class="divider"></div>
            <p>&emsp;&emsp;&emsp;For the bioinformatics project I was working on, I needed to compare the outputs between three guided assemblers: <a href="http://bio-bwa.sourceforge.net/">BWA</a>, <a href="https://github.com/Cibiv/NextGenMap/wiki">NGM</a>, &amp; <a href="http://www.well.ox.ac.uk/project-stampy">STAMPY</a> and three de novo assemblers: <a href="http://soap.genomics.org.cn/SOAPdenovo-Trans.html">SOAPdenovo-Trans</a>, <a href="https://github.com/trinityrnaseq/trinityrnaseq/wiki">Trinity</a>, &amp; <a href="https://www.ebi.ac.uk/~zerbino/velvet/">Velvet</a>. All were used to recreate the exome for 160 different genes, but now I wanted to compare these results to both a reference genome and with each other. Since they were grouped by assembler, I wrote some primer scripts in order to regroup them by individual genes. <a href="RiffleFasta.py">RiffleFasta.py</a> would be used for transposing the assembler grouping to gene symbol.</p>
            <div class="terminal">
              <ol>
                <li><span class="prompt">#Group by genes instead of by assembler</span></li>
                <li><span class="prompt"></span>mkdir Genes</li>
                <li><span class="prompt"></span>python RiffleFasta.py Genes Data/{Anolis_Ref,BWA,NGM,STAMPY,SOAP,Trinity,Velvet}.fasta</li>
                <li><span class="prompt"></span></li>
                <li><span class="prompt">#Create Excel alignments</span></li>
                <li><span class="prompt"></span>mkdir Alignments</li>
                <li><span class="prompt"></span>python Mega.py -o Genes/Anolis $(ls Genes/*.fasta)</li>
                <li><span class="prompt"></span></li>
              </ol>
            </div>
            <p>&emsp;&emsp;&emsp;Finally, I used Mega.py to create 16 Excel files, each containing 10 different alignments. From this, I could easily see that the gap regions shared between all assemblers was due to poor capture efficiency of the enrichment probes used. More interestingly, I learned that STAMPY tended to follow the sequence of the reference genome during reconstruction, leading to numerous nonsensical ambiguities when viewing the protein alignments. Also noteworthy was that none of the assemblers were a clear winner in stitching together contiguous sequences as some would be able to recover areas which others would leave blank. Thus the use of all six (or more) in tandem could be used to create an effective consensus sequence, if one has the sheer computing power and free time (conveniently I have nothing better to do). Check out an example output file: <a href="Cemophora_RSO6ex4_19_LRAT_to_NEUR3.xlsx">Example.xlsx</a>.</p>
            <table>
              <tr>
                <th><a href="../../img/MegaComparison.png"><img src="../../img/MegaComparison.png" alt="Mega Comparison"></a></th>
              </tr>
              <tr>
                <td class="table-title">Protein sequence comparison</td>
              </tr>
            </table>
          </div>
        </div>
      </div>
    </article>
    <!--Sidebar-->
    <aside class="col-sm-3">
      <div class="sidebar" id="spy-target">
        <h3>Mega.py</h3>
        <div class="divider"></div>
        <ul class="nav nav-list">
          <li><a href="#introduction">Introduction</a></li>
          <li><a href="#features">Features</a></li>
          <li><a href="#download">Download</a></li>
          <li><a href="#demonstration">Demonstration</a></li>
        </ul>
      </div>
      <div class="button" id="all-programs">
        <a href="https://github.com/matthewpreston/matthewpreston.github.io/tree/master/programs">All Programs</a>
      </div>
      <div class="round-button" id="top">
        <a href="#">Top</a>
      </div>
    </aside>
  </div>

  <!--Footer-->
  <footer class="container-fluid">
    <p>Updated on August 2016</p>
    <p>Thanks to <a href="http://kellyli.me/">Kelly Li</a> | <a href="http://wummel.github.io/">Wummel</a> | <a href="http://www.danielecook.com/">Daniel E. Cook</a> | <a href="http://www.w3schools.com/">w3schools</a></p>
  </footer>

  <!--JQuery-->
  <script src="../../js/jquery.min.js"></script>
  <script src="../../js/bootstrap.min.js"></script>
  <script src="../../js/article.js"></script>
</body>
</html>
