<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta name="author" content="Matthew Preston">
  <meta name="description" content="">

  <title>AnnotateContigs.py - Annotating de novo Generated Contigs</title>

  <!--CSS-->
  <link rel="stylesheet" href="../../css/bootstrap.min.css">
  <link rel="stylesheet" href="../../css/octicons.min.css">
  <link rel="stylesheet" href="../../css/article.css">
  <link rel="stylesheet" href="../../css/article-python.css">

  <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
  <![endif]-->
</head>
<body data-spy="scroll" data-target="#spy-target" data-offset="60">
  <!--Navbar-->
  <nav class="navbar navbar-default">
    <div class="container">
      <div class="navbar-header">
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#menuNavbar">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../../index.html">Matt Preston</a>
      </div>
      <div class="collapse navbar-collapse" id="menuNavbar">
        <ul class="nav navbar-nav navbar-right">
          <li><a href="../index.html">Programs</a></li>
          <li><a href="../../Resume_Matthew_Preston.pdf" target="_blank">Resume</a></li>
          <li><a href="../../about.html">About</a></li>
        </ul>
      </div>
    </div>
  </nav>

  <div class="row wrapper">
    <!--Main container-->
    <article class="col-sm-9">
      <!--Title-->
      <div class="title">
        <h1>AnnotateContigs.py</h1>
        <div class="small-divider"></div>
        <h4>Annotating <em>de novo</em> Generated Contigs</h4>
      </div>
      <!--Create double border effect-->
      <div class="outer-border">
        <div class="inner-border">
          <!--Presentation Image-->
          <div class="main-img">
            <a href="../../img/AnnotateContigs.svg"><img src="../../img/AnnotateContigs.svg" alt="De Novo Assembly Pipeline"></a>
          </div>
          <!--Introduction-->
          <div class="container-fluid" id="introduction">
            <h3>Introduction</h3>
            <div class="divider"></div>
            <p>&emsp;&emsp;&emsp;A bioinformatics focused script which does what it says on the box: annotates contigs. BLAST your contigs against a subset of NCBI's database and it'll stitch contigs together, effectively reconstructing the original DNA/RNA sequence.</p>
          </div>
          <!--Features-->
          <div class="container-fluid" id="features">
            <h3>Features</h3>
            <div class="divider"></div>
            <ul>
              <li><strong>4 non-mutually exclusive output modes:</strong>
                <ul>
                  <li><strong>Alignments:</strong> Outputs every contig-reference pairwise alignment in one convenient file for each reference gene.
                  </li>
                  <li><strong>Consensus:</strong> Attempts to merge the aligned contigs into one meaningful consensus sequence.
                  </li>
                  <li><strong>Spliced Consensus:</strong> In addition to creating a single consensus sequence, it splices said gene given exon-intron barriers (Note: requires internet access).
                  </li>
                  <li><strong>(Statistical) Data</strong> Provides desirable statistics about the annotation and assembly process, such as % non-coding/coding, breadth of coverage, and % sequence identity.
                  </li>
                </ul>
              </li>
              <li>Multithreading capability</li>
              <li>Even without an explicit set of reference gene, can still attempt annotation of sequences by directly querying NCBI for &quot;sequence hit&quot; metadata (Note: requires BLAST XML file to be generated by aligning to a subset or entire NCBI nt database)</li>
            </ul>
          </div>
          <!--Download/Github/Dependencies-->
          <div class="container-fluid" id="download">
            <h3>Download</h3>
            <div class="divider"></div>
            <table>
              <tr>
                <td><a href="AnnotateContigs.zip" title="Download AnnotateContigs.zip" class="button"><span class="glyphicon glyphicon-compressed"></span> AnnotateContigs.zip</a></td>
                <td><a href="AnnotateContigs.tar.gz" title="Download AnnotateContigs.tar.gz" class="button"><span class="glyphicon glyphicon-compressed"></span> AnnotateContigs.tar.gz</a></td>
                <td><a href="https://github.com/matthewpreston/AnnotateContigs" title="View on GitHub" class="button"><span class="octicon octicon-git-branch"></span>View on GitHub</a></td>
              </tr>
            </table>
          </div>
          <div class="container-fluid" id="dependencies">
            <h3>Dependencies</h3>
            <div class="divider"></div>
            <ul>
              <li><a href="http://biopython.org/">BioPython</a> - For bioinformatical work</li>
              <li><a href="http://xlsxwriter.readthedocs.io/">xlsxwriter</a> - For creating Excel files</li>
            </ul>
          </div>
          <!--Usage/Demonstration-->
          <div class="container-fluid" id="demonstration">
            <h3>Demonstration</h3>
            <div class="divider"></div>
            <p>&emsp;&emsp;&emsp;As my position as bioinformatician for University of Toronto's <a href="https://chang.eeb.utoronto.ca/">Chang Lab</a>, I was tasked with juxtaposing two widely used techniques in genome/transcriptome sequencing: guided and <em>de novo</em> assembly. This script was designed for the latter. Let's cut to the gist; use this if utilizing a set of reference sequences (Note: only FASTA format currently supported):</p>
            <div class="code terminal">
              <ol id="reference">
                <li><span class="prompt"></span>python AnnotateContigs.py \</li>
                <li>&nbsp;&nbsp;-n 10 \</li>
                <li>&nbsp;&nbsp;-r &lt;reference.fasta&gt; \</li>
                <li>&nbsp;&nbsp;-i 0.7 \</li>
                <li>&nbsp;&nbsp;-s 0.7 \</li>
                <li>&nbsp;&nbsp;&lt;contigs.fasta&gt; \</li>
                <li>&nbsp;&nbsp;acds</li>
              </ol>
              <div class="code-btn" data-target="#reference">More</div>
            </div>
            <p>&emsp;&emsp;&emsp;This rejects sequences with sequence identity/similarity less than 70% when compared to those found in reference.fasta (for DNA/RNA, identity and similarity are the same) while outputting sequence alignments, consenses, spliced consenses, and statistical data to a default folder dubbed AnnotateContigs/, all while utilizing a total of 10 threads. If instead of providing a reference database, one can use NCBI's nucleotide database as follows:</p>
            <div class="code terminal">
              <ol id="internet">
                <li><span class="prompt"></span>python AnnotateContigs.py \</li>
                <li>&nbsp;&nbsp;-n 10 \</li>
                <li>&nbsp;&nbsp;-i 0.7 \</li>
                <li>&nbsp;&nbsp;-s 0.7 \</li>
                <li>&nbsp;&nbsp;&lt;contigs.fasta&gt; \</li>
                <li>&nbsp;&nbsp;acds</li>
              </ol>
              <div class="code-btn" data-target="#internet">More</div>
            </div>
            <p>&emsp;&emsp;&emsp;This alternative usage may take a while longer as the program queries NCBI via the internet for each unique sequence encountered in our set of alignments (thus bandwidth can be a concern). Additionally NCBI limits their server access to 3 times a second, which may sound fast to us in human time, but is arduously slow for computers.</p>
            <p>&emsp;&emsp;&emsp;In summary, the contigs will be merged based on the their alignments. This effectively &quot;annotates&quot; the contigs, but allows for the possibiity of polymorphism if there are two contigs that differ in a few sites. Also, if probe tiling was not completely successful due to more variable regions, this program will fill the unaligned regions with gap characters - essentially briding two contigs that would otherwise be separated due to no overlap.</p>
            <p>&emsp;&emsp;&emsp;I must however note for my own conscious's sake: the algorithm here is not foolproof. During its creation, I was naive and eliminated any indels that arose via BLAST alignments. The statistical calculations will be correct, but the consenses *might* be wrong. I duly apologize for this and was working on a more robust algorithm with a proper statistical basis which plagues some of the problems of using this program. Unfortunately, I ran out of time - it's in an intermediate state that requires polish. This new attempt can be seen here: <a href="../RBM/index.html">Reciprocal Best Matching Algorithm</a>. However for similar species, this program as is should be sufficient for biological inference (you'd need an insertion/deletion of a triplet to really trip this up).</p>
          </div>
        </div>
      </div>
    </article>
    <!--Sidebar-->
    <aside class="col-sm-3">
      <div class="sidebar" id="spy-target">
        <a href="#">
          <div class="sidebar_title">
            <h3>AnnotateContigs.py</h3>
          </div>
        </a>
        <div class="divider"></div>
        <ul class="nav nav-list">
          <li><a href="#introduction">Introduction</a></li>
          <li><a href="#features">Features</a></li>
          <li><a href="#download">Download</a></li>
          <li><a href="#demonstration">Demonstration</a></li>
        </ul>
      </div>
      <div class="button" id="all-programs">
        <a href="../index.html">All Programs</a>
      </div>
      <div class="round-button" id="top">
        <a href="#">Top</a>
      </div>
    </aside>
  </div>

  <!--Footer-->
  <footer class="container-fluid">
    <p>Updated on December 2017</p>
    <p>Thanks to Friends &amp; Family | <a href="../../about.html#inspirations">Insiprations</a></p>
  </footer>

  <!--JQuery-->
  <script src="../../js/jquery.min.js"></script>
  <script src="../../js/bootstrap.min.js"></script>
  <script src="../../js/article.js"></script>
</body>
</html>
